File: sdk/node/src/webhooks.ts
import crypto from 'node:crypto';
import { WebhookEventEnvelope } from './types.js';

/**
 * Verify an inbound Fulkruma webhook signature.
 *
 * Header: `Fulkruma-Signature: t=<unix>,v1=<hex>`
 * Payload signed: `${t}.${rawBody}` with HMAC-SHA256 keyed on the
 * endpoint's shared secret.
 *
 * Returns the parsed event envelope on success, throws on bad signature
 * or replay (timestamp older than `toleranceSec`, default 5 min).
 */
export function verifyWebhook<T = unknown>(opts: {
  rawBody: string | Buffer;
  signature: string | undefined | null;
  secret: string;
  toleranceSec?: number;
}): WebhookEventEnvelope<T> {
  if (!opts.signature) throw new Error('missing Fulkruma-Signature header');
  const tolerance = opts.toleranceSec ?? 300;

  const parts: Record<string, string> = {};
  for (const segment of opts.signature.split(',')) {
    const [k, v] = segment.split('=');
    if (k && v) parts[k.trim()] = v.trim();
  }
  const ts = parts.t;
  const v1 = parts.v1;
  if (!ts || !v1) throw new Error('malformed signature header');

  const tsNum = Number.parseInt(ts, 10);
  if (!Number.isFinite(tsNum)) throw new Error('non-numeric timestamp');
  const drift = Math.abs(Math.floor(Date.now() / 1000) - tsNum);
  if (drift > tolerance) throw new Error(`signature timestamp ${drift}s out of tolerance`);

  const bodyStr = typeof opts.rawBody === 'string' ? opts.rawBody : opts.rawBody.toString('utf8');
  const expected = crypto.createHmac('sha256', opts.secret).update(`${ts}.${bodyStr}`).digest('hex');
  const expectedBuf = Buffer.from(expected, 'utf8');
  const givenBuf = Buffer.from(v1, 'utf8');
  if (expectedBuf.length !== givenBuf.length || !crypto.timingSafeEqual(expectedBuf, givenBuf)) {
    throw new Error('bad signature');
  }

  let event: WebhookEventEnvelope<T>;
  try {
    event = JSON.parse(bodyStr) as WebhookEventEnvelope<T>;
  } catch {
    throw new Error('webhook body is not valid JSON');
  }
  return event;
}

File: sdk/go/webhooks.go
package fulkruma

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"strconv"
	"strings"
	"time"
)

// VerifyWebhookOptions tunes VerifyWebhook. A nil pointer uses defaults.
type VerifyWebhookOptions struct {
	// ToleranceSec — max drift in seconds between the signature timestamp
	// and the current clock. Default 300 (5 min). Zero or negative means
	// "use the default".
	ToleranceSec int64
	// Now — clock for the freshness check. Nil means time.Now. Useful in
	// tests.
	Now func() time.Time
}

// VerifyWebhook verifies and parses an inbound Fulkruma webhook.
//
// Fulkruma signs every webhook delivery with::
//
//	Fulkruma-Signature: t=<unix>,v1=<hex>
//
// where `<hex> = HMAC-SHA256(secret, fmt.Sprintf("%s.%s", t, rawBody))`.
// The raw body bytes are what got signed — re-serialising parsed JSON
// will drift whitespace and the signature will never match.
//
// Returns the parsed envelope on success. Returns *Error on any failure
// (missing header, drift, bad signature, malformed JSON).
//
// net/http example:
//
//	func fulkrumaWebhook(w http.ResponseWriter, r *http.Request) {
//	    body, err := io.ReadAll(r.Body)
//	    if err != nil { http.Error(w, "bad body", 400); return }
//	    sig := r.Header.Get("Fulkruma-Signature")
//	    event, err := fulkruma.VerifyWebhook(body, sig, os.Getenv("FULKRUMA_WEBHOOK_SECRET"), nil)
//	    if err != nil { http.Error(w, err.Error(), 400); return }
//	    // switch event.Type { … }; reply 204.
//	}
func VerifyWebhook(rawBody []byte, signatureHeader, secret string, opts *VerifyWebhookOptions) (*WebhookEventEnvelope, error) {
	if signatureHeader == "" {
		return nil, newErr(0, "missing_signature", "missing Fulkruma-Signature header")
	}

	tolerance := int64(300)
	nowFn := time.Now
	if opts != nil {
		if opts.ToleranceSec > 0 {
			tolerance = opts.ToleranceSec
		}
		if opts.Now != nil {
			nowFn = opts.Now
		}
	}

	ts, v1, ok := parseFulkrumaSignature(signatureHeader)
	if !ok {
		return nil, newErr(0, "malformed_signature", "malformed signature header")
	}
	tsNum, err := strconv.ParseInt(ts, 10, 64)
	if err != nil {
		return nil, newErr(0, "malformed_signature", "non-numeric timestamp")
	}
	drift := nowFn().Unix() - tsNum
	if drift < 0 {
		drift = -drift
	}
	if drift > tolerance {
		return nil, newErr(0, "signature_expired",
			"signature timestamp "+strconv.FormatInt(drift, 10)+"s out of tolerance")
	}

	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(ts))
	mac.Write([]byte{'.'})
	mac.Write(rawBody)
	expected := hex.EncodeToString(mac.Sum(nil))

	// Constant-time compare. Fall back to a length check first because
	// hmac.Equal is constant-time only on equal-length inputs.
	givenBytes, err := hex.DecodeString(v1)
	if err != nil || len(givenBytes) != sha256.Size {
		return nil, newErr(0, "bad_signature", "bad signature")
	}
	expectedBytes, _ := hex.DecodeString(expected)
	if !hmac.Equal(expectedBytes, givenBytes) {
		return nil, newErr(0, "bad_signature", "bad signature")
	}

	var env WebhookEventEnvelope
	if err := json.Unmarshal(rawBody, &env); err != nil {
		return nil, newErr(0, "invalid_body", "webhook body is not valid JSON")
	}
	return &env, nil
}

func parseFulkrumaSignature(header string) (ts, v1 string, ok bool) {
	var tSeen, v1Seen bool
	for _, segment := range strings.Split(header, ",") {
		segment = strings.TrimSpace(segment)
		if segment == "" {
			continue
		}
		k, v, found := strings.Cut(segment, "=")
		if !found {
			continue
		}
		switch strings.TrimSpace(k) {
		case "t":
			ts = strings.TrimSpace(v)
			tSeen = true
		case "v1":
			v1 = strings.TrimSpace(v)
			v1Seen = true
		}
	}
	return ts, v1, tSeen && v1Seen
}
